### ISSUE TITLE
Using "import" in JS files doesn't load up latest JS in deployed version of theme 

### ISSUE THREAD

**Issue by GuillermoCasanova** [Shopify/dawn#316]
**Describe the current behavior**

I'm using "import" pattern in my Js files within my assets folder. For example A-File.JS is injected via {{ 'A-File.JS' | asset_url }} within file.liquid -- this loads fine. But if A-File.JS is importing B-FILE.JS, this B-FILE is not the latest version of itself, it's instead pointing to an older version.

When viewing the network tab, assets pulled in via {{asset_url}} have a ?v=8129822768428393845 at the end. Whereas my imported JS files via JS files do not. This  seems likely to be the issue?

When developing locally things work just fine, but when the theme is pushed to the store and viewed via the link, things break. Even imports of images via JS files don't work. Even though looking at the code for the theme via "Edit Code", everything is latest. 

**Describe the expected behavior**
Using import should pull up latest JS file. Shopify even recommends using "import" strategy when optimizing JS code. 

**Version information (Dawn, browsers and operating systems)**
- Dawn Version: 1.0.0
- Chrome Version 91.0.4472.114
- macOS Version 11.3.1

**Additional context/screenshots**

"CS-COMING-SOON.JS" is importing the three js files under them. 
![Screen Shot 2021-08-04 at 12 17 57 PM](https://user-images.githubusercontent.com/9218601/128217117-7979df08-9e6a-4456-883b-a69abeb88d42.png)

**Comment by olafghanizadeh** [comment]
> When viewing the network tab, assets pulled in via {{asset_url}} have a ?v=8129822768428393845 at the end. Whereas my imported JS files via JS files do not. This seems likely to be the issue?

This is for cache busting purposes, so that if a JS asset is changed and a user with a cached version visits the site, the browser is instructed to ignore what is cached and load the most recent version. I do not think this is the issue.

Can you share the code of how you are importing your JS files?

**Comment by GuillermoCasanova** [comment]
Yeah that makes sense and is super useful for cache, seems to only be working with files pulled with {{<FILE> | asset_url}} though

This is how JS files are being imported (example of js files and images)

THREE is a module from CDNJS I downloaded, file is in the assets folder
cs-first-person-controls.js is a class I  made, also in assets folder  

the other tree are images, these are also not importing, so I'm having to have the image urls in the HTML and pull from that in the JS instead of just importing the images being used in the JS. 

```
import * as THREE from './three.module.js';
import FirstPersonControls  from './cs-first-person-controls-custom.js';

let glow = './assets/glow.png';
let starSprite = './assets/glow_small.png';
let starmap = './assets/stars.png';
```

**Comment by olafghanizadeh** [comment]
I don’t think you can import JS like that from a JS file on Shopify’s
servers. It does not have access to the file system the same way you do on
your local machine,  and it would need the full asset URL.

In Dawn the JS files are split into multiple files and loaded when
necessary.

If your plan is to rely on some library you might need to use a build
pipeline to compile a bundled version.

**Comment by GuillermoCasanova** [comment]
Would it be fair to say then that Shopify's advice on "import on usage" implies it assumes a bundler needs to be used? 

https://shopify.dev/themes/best-practices/performance#load-non-critical-resources-on-interaction

But I gotcha on that then. 

Atm there is no bundler that works from what I have found (Shopify Packer, which I use for older pre-dawn themes, is incompatible with new JSON templates). I also don't know how to create a dev environment  + build pipeline pipeline from scratch that also uses the Shopify CLI. 

Shopify CLI is fantastic for running things so I would love to leverage that, but maybe for now it's a matter of developing via the Shopify CLI, thenusing Webpack as a bundler to bundle into a /dist and then pushing that directory with github?

My experience is limited in that realm, so I think I'll be forced to stick to using Shopify's way / importing my modules this way temporarily until someone makes a build pipeline preferably with Webpack. 

However, any direction is welcome! Otherwise we can consider this closed then. 

Thank you so much!

**Comment by GuillermoCasanova** [comment]
Ah thank you @olafghanizadeh ! This helps a ton in grasping this. 

I think I get the idea of doing imports, but doing em' via the Liquid code, etc.

I'll have to look into Github Actions, not too familiar myself as I've been used to using Webpack instead. Thanks!

**Comment by eballeste** [comment]
@GuillermoCasanova i spent my whole day breaking my head trying to solve the same issue! 

what you are describing are ESM module imports and you can absolutely load javascript files within other javscript modules as in your example since they are all loading relative to the `assets` folder. The problem is that the file is being aggressively cached by your browser!

To solve this, Shopify will append a random querystring to the end of a script any time you upload an updated script file by the use of their liquid `asset_url` filter. However, when you do an import from within a javascript file, the browser does not have access to Shopify's cache busting mechanism so it will always load the older cached version you are seeing without the hash at the end of the file name. 

The solution, is that you have to cache bust your internal ESM imports. There's no way around this atm. 
You can do this from within your JS files which works but is cumbersome because you will have to manually add a new querystring to the end of your imports everytime you push a new update:

```javascript
const Foo = import("./js-foo.js?v=999";
```

or you can always force a cache-bust:
```javascript
const Foo = import("./js-foo.js?v=" + (Date.now() / 1000));
```

or you can use an import map! This will allow you to cache bust your internal imports using Shopify's cache-busting mechanism through the use of the `asset_url` filter.

```liquid
<script type="importmap">
  {
    "imports": {
      "foofoo": "{{ 'foo.js' | asset_url }}", // this will make sure your imports are always cache-busted
      "barbar": "https://example.com/examples/bar.js"
    }
  }
</script>

<script type="module">
  import Foo from 'foofoo'
  import Bar from 'barbar'
</script>
```

here's a good resource on [using ESM import maps](https://blog.logrocket.com/es-modules-in-browsers-with-import-maps/)

**Comment by eballeste** [comment]
just a quick follow-up, support for import maps started just last month in Safari (16.4) browsers which not so many people have. the ES Module Shim works, but is clunky and ruins certain page load experiences so i though about this a little more and came up with a solution that both cache-busts and allows you to use something similar to an import map. it also isn't restricted to a single use and can even be used within section liquid templates. 

the trick is to use a simple object to act like the import map and instead of using static imports, you use dynamic imports (which [has much more support](https://caniuse.com/es6-module-dynamic-import) than importmaps including older safari browsers)

`layout.liquid`
```html
<script type="module">
  import initFooBar from '{{ 'js-foobar.js' | asset_url }}';
  
  const importmap = {
    foo: '{{ 'js-foo.js' | asset_url }}',
    bar: '{{ 'js-bar.js' | asset_url }}',
  };
  
  initFooBar(importmap);
</script>
```

`js-foobar.js`
 ```javascript
 const initFooBar = async function(scripts) {
   let Foo = await import(scripts.foo); // cache busted import thanks to asset_url filter
   let { Bar, BarBar } = await import(scripts.bar); // cache busted import thanks to asset_url filter
   
   const initScripts = function() {
     Foo();
     Bar();
     BarBar();
   };
   
   // async dynamic imports cause race conditions so we first check if still loading
   if (document.readyState !== 'loading') {
    initScripts();
  } else {
    document.addEventListener('DOMContentLoaded', function() {
      initScripts();
    }, false);
  }
 };
 ```

**Comment by eballeste** [comment]
@GriffinJohnston i'm not sure if Shopify will be able to do something about it, this seems to be a reoccurring issue with browsers and [ES module imports](https://stackoverflow.com/questions/47675549/how-do-i-cache-bust-imported-modules-in-es6) in general. 

In my last example the scripts are not polluting the global namespace they are being scoped to variables in the `initFoobar` method which is the same exact thing `<script type="importmaps"></script>` seems to be solving, a way to control import scripts by using aliases.

**Comment by devinenoise** [comment]
I found that using <script type="module" src="{{ 'script-name.js' | asset_url | split: '?' | first  }}"></script> busts the cache and only loads a single version instead of one with the version number and one without.
